src/protocol.py

```diff
@@ -105,6 +105,29 @@ def networkType(host):
     return 'IPv6'
 
 
+def network_group(host):
+    """Canonical identifier of network group
+       simplified, borrowed from
+       GetGroup() in src/netaddresses.cpp in bitcoin core"""
+    network_type = networkType(host)
```

Comment:
```
2019-11-16 13:44:27,808 - CRITICAL - Unhandled exception
Traceback (most recent call last):
  File "/usr/lib/python-exec/python2.7/pybitmessage", line 11, in <module>
    execfile(script_file, new_globals)
  File "/home/glitch/test/PyBitmessage/pybitmessage/bitmessagemain.py", line 505, in <module>
    main()
  File "/home/glitch/test/PyBitmessage/pybitmessage/bitmessagemain.py", line 501, in main
    mainprogram.start()
  File "/home/glitch/test/PyBitmessage/pybitmessage/bitmessagemain.py", line 317, in start
    objectProcessorThread = objectProcessor()
  File "/home/glitch/test/PyBitmessage/pybitmessage/class_objectProcessor.py", line 57, in __init__
    self._ack_obj = bmproto.BMStringParser()
  File "/home/glitch/test/PyBitmessage/pybitmessage/network/bmproto.py", line 692, in __init__
    super(BMStringParser, self).__init__()
  File "/home/glitch/test/PyBitmessage/pybitmessage/network/bmproto.py", line 75, in __init__
    self.network_group = protocol.network_group(address)
  File "/home/glitch/test/PyBitmessage/pybitmessage/protocol.py", line 112, in network_group
    network_type = networkType(host)
  File "/home/glitch/test/PyBitmessage/pybitmessage/protocol.py", line 101, in networkType
    if host.find('.onion') > -1:
AttributeError: 'NoneType' object has no attribute 'find'
```

Maybe because of no ipv6 support in python.

Reply:
Well, it maybe mine and fixes with this patch:
```patch
diff --git a/src/network/bmproto.py b/src/network/bmproto.py
index 06313888..5ba1ad0e 100644
--- a/src/network/bmproto.py
+++ b/src/network/bmproto.py
@@ -65,7 +65,8 @@ class BMProto(AdvancedDispatcher, ObjectTracker):
     maxTimeOffset = 3600
     timeOffsetWrongCount = 0
 
-    def __init__(self, address=None, sock=None):    # pylint: disable=unused-argument, super-init-not-called
+    def __init__(self, address=None, sock=None):
+        # pylint: disable=unused-argument, super-init-not-called
         AdvancedDispatcher.__init__(self, sock)
         self.isOutbound = False
         # packet/connection from a local IP
@@ -689,7 +690,7 @@ class BMStringParser(BMProto):
     A special case of BMProto used by objectProcessor to send ACK
     """
     def __init__(self):
-        super(BMStringParser, self).__init__()
+        super(BMStringParser, self).__init__('127.0.0.1')
         self.destination = Peer('127.0.0.1', 8444)
         self.payload = None
         ObjectTracker.__init__(self)
```

Reply:
But now I have exception in Asyncore:
```
Exception in thread Asyncore:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/home/glitch/test/PyBitmessage/pybitmessage/network/networkthread.py", line 15, in run
    BMConnectionPool().loop()
  File "/home/glitch/test/PyBitmessage/pybitmessage/network/connectionpool.py", line 319, in loop
    self.addConnection(Socks5BMConnection(chosen))
  File "/home/glitch/test/PyBitmessage/pybitmessage/network/tcp.py", line 290, in __init__
    TCPConnection.__init__(self, address=address, sock=self.socket)
  File "/home/glitch/test/PyBitmessage/pybitmessage/network/tcp.py", line 47, in __init__
    BMProto.__init__(self, address=address, sock=sock)
  File "/home/glitch/test/PyBitmessage/pybitmessage/network/bmproto.py", line 76, in __init__
    self.network_group = protocol.network_group(address)
  File "/home/glitch/test/PyBitmessage/pybitmessage/protocol.py", line 122, in network_group
    network_type = networkType(host)
  File "/home/glitch/test/PyBitmessage/pybitmessage/protocol.py", line 111, in networkType
    if host.find('.onion') > -1:
AttributeError: 'Peer' object has no attribute 'find'


```

Reply:
c74fda1

Reply:
Since this is only checked for outbound connections (reasoning explained in the bitcoin source code), I think it should be addressed elsewhere (see updated commits).

Reply:
Added check for None as a workaround.